src/extensions/idle.rs

```diff
@@ -91,24 +100,28 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     /// Internal helper that doesn't consume self.
     ///
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
-    fn wait_inner(&mut self) -> Result<()> {
+    fn wait_inner(&mut self, reconnect: bool) -> Result<WaitOutcome> {
         let mut v = Vec::new();
         match self.session.readline(&mut v).map(|_| ()) {
             Err(Error::Io(ref e))
                 if e.kind() == io::ErrorKind::TimedOut || e.kind() == io::ErrorKind::WouldBlock =>
             {
-                // we need to refresh the IDLE connection
-                self.terminate()?;
-                self.init()?;
-                self.wait_inner()
+                if reconnect {
+                    self.terminate()?;
+                    self.init()?;
+                    self.wait_inner(reconnect)
+                } else {
+                    Ok(WaitOutcome::TimedOut)
```

Comment:
Thanks for the patch -- it looks pretty reasonable to me! I think we can avoid changing the public API though (and thus avoid a major version bump). What if we just returned the underlying error when `reconnect == false`? That way, the caller can determined if a timeout occurred simply by inspecting the type of the error. We should of course document that in the documentation for `wait_timeout`. And that way, the return type for all the `wait` methods can stay the same as they were.

Reply:
Makes sense, although using an error condition to signal the very purpose of the function feels a bit suboptimal.

Especially because clients would need to understand the semantics of the underlying session and (if my experiments are correct) would have to check (at least) for  `WouldBlock` (which sounds counter-intuitive and I don't know what that value means :) )

Also, it would be nice to integrate getting back *what* was changed from the `Handle` (there's another open issue for that, right?), in which case what I called `WaitOutcome` would be better suited to turn into something more "dataful".

So anyway, yes, maybe the trouble of a major version bump could wait for the integration of that part too?

Up to you :)

Reply:
I think the `WouldBlock` error code should actually not be handled there at all, and `TimedOut` is the only one that "matters". But you're right, we could translate that particular error for `wait_timeout` specifically. I don't think we need to change the other methods though.

I agree with you that `WaitOutcome` would be a good addition eventually when we also add data, but probably not worth a breaking change until we do. For the time being, how about:

 - Revert the return type of all the `pub` methods
 - Mark `wait_timeout` as `#[deprecated]`
 - Add `pub fn wait_with_timeout`, which returns `Result<WaitOutcome>`

Reply:
> I think the WouldBlock error code should actually not be handled there at all, and TimedOut is the only one that "matters"

That was my first attempt: a new branch in the `match`, on `Err(e) if e.kind() == io::ErrorKind::TimedOut`. But after the timeout the function did not return :/

> For the time being, how about:

That makes a lot of sense

Reply:
Sorry it took me a while :P

Is this what you meant?

Reply:
Yes, exactly!